sb3_contrib/ars/ars.py

```diff
@@ -0,0 +1,276 @@
+
+        if n_top is None:
+            n_top = n_delta
+        self.n_top = n_top
+
+        self.n_workers = None  # We check this at training time ... I guess??
+
+        if policy_kwargs is None:
+            policy_kwargs = {}
+        self.policy_kwargs = policy_kwargs
+
+        self.seed = seed
+        self.alive_bonus_offset = alive_bonus_offset
+
+        self.zero_policy = zero_policy
+        self.theta = None  # Need to call init model to initialize theta
+
+        if (
+            _init_setup_model
+        ):  # TODO ... what do I do if this is false? am i guaranteed that someone will call this before training?
+            self._setup_model()
+
+    @classmethod  # Override just to change the default device argument to "cpu"
+    def load(
+        cls,
+        path: Union[str, pathlib.Path, io.BufferedIOBase],
+        env: Optional[GymEnv] = None,
+        device: Union[th.device, str] = "cpu",
+        custom_objects: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> "BaseAlgorithm":
+        return super().load(path, env, device, custom_objects, **kwargs)
+
+    def _setup_model(self) -> None:
+        self.set_random_seed(self.seed)
+        self.rng = np.random.default_rng(self.seed)
+
+        self.policy = self.policy_class(self.observation_space, self.action_space, **self.policy_kwargs)
+        self.dtype = self.policy.parameters().__next__().dtype  # This seems sort of like a hack
+        self.theta = th.nn.utils.parameters_to_vector(self.policy.parameters()).detach().numpy()
+
+        if self.zero_policy:
+            self.theta = np.zeros_like(self.theta)
+            theta_tensor = th.tensor(self.theta, requires_grad=False, dtype=self.dtype)
+            th.nn.utils.vector_to_parameters(theta_tensor, self.policy.parameters())
+
+        self.n_params = len(self.theta)
+        self.policy = self.policy.to(self.device)
+
+    def _collect_rollouts(self, policy_deltas, callback):
+        with th.no_grad():
+            batch_steps = 0
+            theta_idx = 0
+
+            # Generate 2*n_delta candidate policies by adding noise to the current theta
+            candidate_thetas = np.concatenate([self.theta + policy_deltas, self.theta - policy_deltas])
+            candidate_returns = np.zeros(candidate_thetas.shape[0])  # returns == sum of rewards
+            self.ep_info_buffer = []
+
+            callback.on_rollout_start()
+            while theta_idx < candidate_returns.shape[0]:
+                policy_list = []
+
+                # We are using a vecenv with n_envs==n_workers. We batch our candidate theta evaluations into vectors
+                # of length n_workers
+                for _ in range(self.n_workers):
+                    theta_tensor = th.tensor(candidate_thetas[theta_idx], dtype=self.dtype)
+                    th.nn.utils.vector_to_parameters(theta_tensor, self.policy.parameters())
+                    policy_list.append(copy.deepcopy(self.policy))
```

Comment:
Hmm this is probably a big bottleneck. Wonder if you could create a population of N whose parameters you keep changing (or does population size change?)

Reply:
during profiling I saw that almost all the time is spent in env.step or policy.predict. ARS typically uses linear policies so the amount of data here is actually fairly small. preallocating the list would help a bit, perhaps just initializing with policy_list =[copy.deepcopy(policy) for _ in range(n_workers)], IIRC list comprehensions avoid the extra copies that append can run into.